### DISCUSSION TITLE
i-see Sensor

### DISCUSSION THREAD

**Discussion by umtauscher** [echavet/MitsubishiCN105ESPHome#64] [Q&A] [answered]
Hi,
greate work!
I have just got my MSZ-LN35 working with this. Wow, bye bye Melcloud, finally!

On problem though:
The i-see sensor is always showing off in my setup. Even when I use the IR remote I cannot see if it is switched on or off.
Am I missing something? How is it supposed to work.
Anybody with a bit more knowlage than me, could you please help?
TIA
Wilhelm

**Comment by echavet** [comment]
Hi Wilhelm,
Thank you. 
Actually I have kept the i-see feature from the original swicago lib but I've never been able to test it because my units don't support this feature.

My understanding is that the function should be activated from the IR remote control. Because in my component it is a read-only sensor.
Then the unit should report a state indicating someone has been detected. 
But not sure about that.

Does the melcloud support this sensor ?

**Comment by umtauscher** [comment]
No, Melcloud doesn‘t support this. 
I am unable to see any change in HA/CN105 even when I activate it with the IR Remote.

**Comment by umtauscher** [comment]
I just checked:
CN105ESPHome definately switches the i-see sensor off. 
I just tested with an unmodified unit.
When using the CN105ESPHome interface I cannot switch the i-see sensor on.
So it seems there is a setting that would switch via cn105. I have another unit where I installed Mitsubishi2MQTT. There the i-see sensor works normally
So I hope you can find the difference between the 2 implementations and make it to change the sensor state via CN105ESPHome

**Reply by echavet** [reply]
Thank you.
Great info. I'm going to compare...

I opened an issue: https://github.com/echavet/MitsubishiCN105ESPHome/issues/65

**Comment by umtauscher** [comment]
Ok, I have to correct myself somewhat.
Today I powercycled the units to get a bit more consistent results.
After the powercycle the first I did was to user the IR Remote to activate the I-See sensor and guess what: It worked.
I hhave not yet found out which combination of the settings of your software switched it off. For now it's working just like it should.

Is there any way to find out additional commands of the CN105 interface e.g. by reading the logs of sniffing on the status changes the IR remote sets?
Thanks

**Comment by echavet** [comment]
Well Yes and... no,

Here is the log settings I use to debug:

```yaml

logger:    
  level: DEBUG
  logs:
    EVT_SETS : DEBUG
    WIFI : INFO
    MQTT : INFO
    WRITE_SETTINGS : DEBUG
    SETTINGS : DEBUG
    STATUS : INFO
    CN105Climate: WARN
    CN105: INFO
    climate: WARN
    sensor: WARN
    chkSum : INFO
    WRITE : DEBUG
    READ : DEBUG
    Update-ACK : ERROR
    Header: INFO
    Decoder : INFO
    CONTROL_WANTED_SETTINGS: INFO    
    testMutex: DEBUG
    UPDT_ITVL: DEBUG
    CYCLE: DEBUG
```

You should see all the SEND and RECEIVE packets.

But, actually, as far as I remember, there is absolutely no "send packet" containing anything about isee sensor.

In the file hp_wrintings.cpp, you will find all the functions and methods sending commands to the unit. 
Most of them are static packets (FirstConnection packet, Infos packet...). 

And one of them is dynamic, this is the commands one:
 packet are build within the ``void CN105Climate::createPacket(uint8_t* packet) {`` method.

This is common with all the implementations you will find because they all use the SwiCago lib. I don't really use it, but I have integrated the lib in my code. So indirectly I do. 

packet are prepared with a header:
```
static const uint8_t HEADER[HEADER_LEN] = { 0xfc, 0x41, 0x01, 0x30, 0x10, 0x01, 0x00, 0x00 };
```
Then packet are filled up with wanted settings:

<details>
<summary>Here is the detailed code (no iSee data)</summary>

```cpp

void CN105Climate::createPacket(uint8_t* packet) {
    prepareSetPacket(packet, PACKET_LEN);

    //ESP_LOGD(TAG, "checking differences bw asked settings and current ones...");
    ESP_LOGD(TAG, "building packet for writing...");

    if (this->wantedSettings.power != nullptr) {
        ESP_LOGD(TAG, "power -> %s", getPowerSetting());
        packet[8] = POWER[lookupByteMapIndex(POWER_MAP, 2, getPowerSetting(), "power (write)")];
        packet[6] += CONTROL_PACKET_1[0];
    }

    if (this->wantedSettings.mode != nullptr) {
        ESP_LOGD(TAG, "heatpump mode -> %s", getModeSetting());
        packet[9] = MODE[lookupByteMapIndex(MODE_MAP, 5, getModeSetting(), "mode (write)")];
        packet[6] += CONTROL_PACKET_1[1];
    }

    if (wantedSettings.temperature != -1) {
        if (!tempMode) {
            ESP_LOGD(TAG, "temperature (tempmode is false) -> %f", getTemperatureSetting());
            packet[10] = TEMP[lookupByteMapIndex(TEMP_MAP, 16, getTemperatureSetting(), "temperature (write)")];
            packet[6] += CONTROL_PACKET_1[2];
        } else {
            ESP_LOGD(TAG, "temperature (tempmode is true) -> %f", getTemperatureSetting());
            float temp = (getTemperatureSetting() * 2) + 128;
            packet[19] = (int)temp;
            packet[6] += CONTROL_PACKET_1[2];
        }
    }

    if (this->wantedSettings.fan != nullptr) {
        ESP_LOGD(TAG, "heatpump fan -> %s", getFanSpeedSetting());
        packet[11] = FAN[lookupByteMapIndex(FAN_MAP, 6, getFanSpeedSetting(), "fan (write)")];
        packet[6] += CONTROL_PACKET_1[3];
    }

    if (this->wantedSettings.vane != nullptr) {
        ESP_LOGD(TAG, "heatpump vane -> %s", getVaneSetting());
        packet[12] = VANE[lookupByteMapIndex(VANE_MAP, 7, getVaneSetting(), "vane (write)")];
        packet[6] += CONTROL_PACKET_1[4];
    }

    if (this->wantedSettings.wideVane != nullptr) {
        ESP_LOGD(TAG, "heatpump widevane -> %s", getWideVaneSetting());
        packet[18] = WIDEVANE[lookupByteMapIndex(WIDEVANE_MAP, 7, getWideVaneSetting(), "wideVane (write)")] | (wideVaneAdj ? 0x80 : 0x00);
        packet[7] += CONTROL_PACKET_2[0];
    }


    // add the checksum
    uint8_t chkSum = checkSum(packet, 21);
    packet[21] = chkSum;
    
}
```
</details>

<details>
<summary>Here is the reading part, where isee data is read: </summary>

```cpp
void CN105Climate::getSettingsFromResponsePacket() {
    heatpumpSettings receivedSettings{};
    ESP_LOGD("Decoder", "[0x02 is settings]");
    //02 00 00 01 08 0A 00 07 00 00 03 AA 00 00 00 00 94
    //this->last_received_packet_sensor->publish_state("0x62-> 0x02: Data -> Settings");        
    receivedSettings.connected = true;      // we're here so we're connected (actually not used property)
    receivedSettings.power = lookupByteMapValue(POWER_MAP, POWER, 2, data[3], "power reading");
    receivedSettings.iSee = data[4] > 0x08 ? true : false;
    receivedSettings.mode = lookupByteMapValue(MODE_MAP, MODE, 5, receivedSettings.iSee ? (data[4] - 0x08) : data[4], "mode reading");

    ESP_LOGD("Decoder", "[Power : %s]", receivedSettings.power);
    ESP_LOGD("Decoder", "[iSee  : %d]", receivedSettings.iSee);
    ESP_LOGD("Decoder", "[Mode  : %s]", receivedSettings.mode);

    if (data[11] != 0x00) {
        int temp = data[11];
        temp -= 128;
        receivedSettings.temperature = (float)temp / 2;
        this->tempMode = true;
        //ESP_LOGD("Decoder", "tempMode is true");
    } else {
        receivedSettings.temperature = lookupByteMapValue(TEMP_MAP, TEMP, 16, data[5], "temperature reading");
    }

    ESP_LOGD("Decoder", "[Temp °C: %f]", receivedSettings.temperature);

    receivedSettings.fan = lookupByteMapValue(FAN_MAP, FAN, 6, data[6], "fan reading");
    ESP_LOGD("Decoder", "[Fan: %s]", receivedSettings.fan);

    receivedSettings.vane = lookupByteMapValue(VANE_MAP, VANE, 7, data[7], "vane reading");
    ESP_LOGD("Decoder", "[Vane: %s]", receivedSettings.vane);


    receivedSettings.wideVane = lookupByteMapValue(WIDEVANE_MAP, WIDEVANE, 7, data[10] & 0x0F, "wideVane reading", WIDEVANE_MAP[2]);
    wideVaneAdj = (data[10] & 0xF0) == 0x80 ? true : false;

    /*if ((data[10] != 0) && (this->traits_.supports_swing_mode(climate::CLIMATE_SWING_HORIZONTAL))) {    // wideVane is not always supported
        receivedSettings.wideVane = lookupByteMapValue(WIDEVANE_MAP, WIDEVANE, 7, data[10] & 0x0F, "wideVane reading");
        wideVaneAdj = (data[10] & 0xF0) == 0x80 ? true : false;
        ESP_LOGD("Decoder", "[wideVane: %s (adj:%d)]", receivedSettings.wideVane, wideVaneAdj);
    } else {
        ESP_LOGD("Decoder", "widevane is not supported");
    }*/

    if (this->iSee_sensor_ != nullptr) {
        this->iSee_sensor_->publish_state(receivedSettings.iSee);
    }

    this->heatpumpUpdate(receivedSettings);

}
```
</details>

**Comment by rudizl** [comment]
My HP is FH50VE and ISEE is always "ON". Will make some test this evening at home.

**Comment by umtauscher** [comment]
Hi,
what Eric wrote is a bit beyond my abilities...
Nevertheless, there seems to be some settings combination that switches the i-See off. When that happens, it seems I cannot switch it on again, even with the IR-remode.

**Comment by umtauscher** [comment]
And just some other observations:
Melcloud also has an auto setting for the horizontal vane. 
This is missing in the UI
Perhaps it has something to do with this.

**Comment by umtauscher** [comment]
Hi Eric,
and since you asked, here is what can be set via Melcloud:
![Bildschirmfoto 2024-03-21 um 14 45 36](https://github.com/echavet/MitsubishiCN105ESPHome/assets/24495410/5b92a7ff-57bd-4b10-9c27-a4ec06939839)

**Comment by umtauscher** [comment]
I still can't firgure out, how I can switch the i-See sensor on or off with the IR-Remote. I thought I had it, but today all theory went out of the windows again - one of the units has it switched on, the other off and I can't switch it on again.

**Comment by umtauscher** [comment]
And short after I wrote that, I found an Action setting in the Melcloud app which can restrict certain settings for the IR remote (local controller). The unit in question had the "operation mode" set to disabled, which led to the i-See sensor stuck in the off position. After I changed that setting, I was able to change the i-see sensor for that unit with the remote again.

So still it would be nice to figure out how to control it from ESPhome and how to get the auto setting for the horizontal vane.

**Comment by disruptivepatternmaterial** [comment, marked as answer]
I have some units with iSee and what I see is that the sensor is on when I am in the room and off when I am not, so that would seem to suggest there are 2 bits here, 1 for the current status (someone there or not) and the second to tell the unit to use it or not! My problem is that the ones I am using for testing using my logic board are easy to get to. The ones with iSee are not...so that will be a lot more work, but I will get to it.

**Comment by umtauscher** [comment]
Maybe this only happens, when you set the home icon in the remote too?
